### PR TITLE
Signup: DSS: Add DangerousHtml component to prevent rendering errors

### LINKED ISSUE #1890: Signup: DSS: Returning to DSS using Back link causes errors
Steps to reproduce:

1. Visit the DSS flow, eg: http://calypso.localhost:3000/start/dss/
2. Click on a theme; Signup will move to step 2.
3. Click the "Back" link at the bottom of the page to return to step 1.
4. Errors!

screen shot 2015-12-21 at 1 55 12 pm

### PR DESCRIPTION
**Pull request by sirbrillig**
Normally it's possible to use the special `dangerouslySetInnerHTML` prop of a
component to add raw HTML into that component, but the markup is injected as-is,
which means that any invalid markup will risk breaking React's virtual DOM.
(Generally this results in errors like `Invariant Violation: Expected markup to
render 15 nodes, but rendered 10`).

DangerousHtml creates a tag (of the tag name matching `tag`) and uses the web
API's `innerHTML` to inject the HTML into the tag rather than using
`dangerouslySetInnerHTML`. Using JavaScript to set the HTML causes the browser
to fix any invalid markup and prevents it from breaking the renderer.

Fixes #1890 

## Testing

1. Visit the DSS flow, eg: http://calypso.localhost:3000/start/dss/
2. Click on a theme; Signup will move to step 2.
3. Click the "Back" link at the bottom of the page to return to step 1.
4. Verify the DSS page displays correctly (no sign of weird rendering).
5. Click on another theme.
6. Complete the signup flow.
7. Verify that the site is created and has the theme you chose.

### CODE REVIEW

File: client/components/dangerousHtml/index.jsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
+import ReactDom from 'react-dom';
+
+export default React.createClass( {
+	propTypes: {
+		tag: React.PropTypes.string.isRequired,
```
Comment (aduth) [inline client/components/dangerousHtml/index.jsx]: Can we make this optional and default to `div` ?

Comment (sirbrillig) [inline client/components/dangerousHtml/index.jsx]: That's a good idea. I actually kind of thought of that when I was writing the README.

File: client/components/dangerousHtml/index.jsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+	componentDidUpdate( prevProps ) {
+		if ( prevProps.html !== this.props.html ) {
+			this.componentDidMount();
+		}
+	},
+
+	render() {
+		var tag = this.props.tag;
+		var props = {
```
Comment (aduth) [inline client/components/dangerousHtml/index.jsx]: Seems to be a good case to simply pass along any inherited props, except the ones that you have custom behavior for, e.g.:

```jsx
return React.createElement( tag, omit( this.props, 'tag', 'html ) );
```

(Means you could also lose the `className` `propType`)

File: client/components/dangerousHtml/index.jsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+		className: React.PropTypes.string,
+	},
+
+	componentDidMount() {
+		ReactDom.findDOMNode( this ).innerHTML = this.props.html;
+	},
+
+	componentDidUpdate( prevProps ) {
+		if ( prevProps.html !== this.props.html ) {
+			this.componentDidMount();
```
Comment (aduth) [inline client/components/dangerousHtml/index.jsx]: We shouldn't call a lifecycle method like this. Instead, extract the logic to a function and call it from both places:

```jsx
componentDidMount() {
	this.updateHtml();
},

componentDidUpdate( prevProps ) {
	if ( prevProps.html !== this.props.html ) {
		this.updateHtml();
	}
},

updateHtml() {
	ReactDom.findDOMNode( this ).innerHTML = this.props.html;
}
```

File: client/components/dangerousHtml/index.jsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
```
Comment (aduth) [inline client/components/dangerousHtml/index.jsx]: Should have `External dependencies` docblock.

File: client/components/dangerousHtml/index.jsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import ReactDom from 'react-dom';
+
+export default React.createClass( {
+	propTypes: {
+		tag: React.PropTypes.string.isRequired,
+		html: React.PropTypes.string.isRequired,
```
Comment (aduth) [inline client/components/dangerousHtml/index.jsx]: We could probably default to an empty string.

File: client/signup/steps/dss/screenshot.jsx [hunk 6]
```diff
@@ -4,6 +4,8 @@
 import React from 'react';
 import classnames from 'classnames';
 
+import DangerousHtml from 'components/danger';
```
Comment (aduth) [inline client/signup/steps/dss/screenshot.jsx]: Bad import? Shouldn't this be `components/dangerous-html` ?

File: client/components/dangerousHtml/index.jsx (hunk 4 above)
Comment (aduth) [inline client/components/dangerousHtml/index.jsx]: Directory name should be `dangerous-html`, not `dangerousHtml`

**Comment by ehg** [comment]
I'm concerned that setting `innerHTML` directly will break React's tree conciliation pretty badly, which I've seen cause browser crashes before. Also, it's completely incompatible with server side rendering :(

File: client/signup/steps/dss/screenshot.jsx (hunk 6 above)
Comment (sirbrillig) [inline client/signup/steps/dss/screenshot.jsx]: Oops. Yes. I changed the name and forgot to change this.

**Comment by aduth** [comment]
Discussed a bit privately, but let me reiterate here that anything related to dangerous HTML throws red flags in my book and should be avoided at all costs. Given that, I understand the need to be a bit more forgiving with the markup here.

I had mentioned the idea of using an `<iframe />` instead of dangerous HTML, and I think if this is possible, we should try to do so. See the [`<Shortcode />` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/shortcode) as an example, which we could probably abstract to a `<SandboxedIframe />` component for rendering arbitrary markup. Specifically, see [`<ShortcodeFrame />`](https://github.com/Automattic/wp-calypso/blob/master/client/components/shortcode/frame.jsx).

Also, should we consider adopting similar mechanisms as used by Facebook to deter usage of this component? A verbose component name or prop indicating that it's poor-form to use?

**Comment by sirbrillig** [comment]
> A verbose component name

Agreed, that was my intention with the name `DangerousHtml` ;-) But I'm open to other ideas.

**Comment by sirbrillig** [comment]
> Also, it's completely incompatible with server side rendering

Hm. That's a good point. The other idea I had was to use some other module to fix invalid HTML. I'll look to see if something like `cheerio` can do it.

**Comment by aduth** [comment]
>But I'm open to other ideas.

Maybe `dangerouslySetInnerHtml` as the prop name instead of `html` ? Redundant, yes, but that's kind of the point :smile:

**Comment by ehg** [comment]
> Hm. That's a good point. The other idea I had was to use some other module to fix invalid HTML. 

We could do this on the server side, and store the valid markup in the theme CPT/ElasticSearch.

**Comment by sirbrillig** [comment]
> Maybe dangerouslySetInnerHtml as the prop name

I changed it to `dangerousHtml` since I probably can't use the already-special `dangerouslySetInnerHtml`. I also added this to the README: 

**Warning**: This is still just as risky as using `dangerouslySetInnerHTML`
directly. Please avoid this component if at all possible.

**Comment by sirbrillig** [comment]
Trying a different approach in #1903 by fixing the markup before it is rendered.

**Comment by nylen** [comment]
> Please avoid this component if at all possible.

I agree with the README here :D

`dangerouslySetInnerHTML` is discouraged for good reasons, and I'd rather see us getting rid of existing uses than adding more ways to use it.

A couple of alternative approaches being discussed at #1903.

**Pull request closed without merging**